docs: :memo: add check and issue guide (#47)
seedcase-project/check-datapackage: docs/guide/check.qmd

---
title: "Checking a Data Package descriptor"
jupyter: python3
# TODO: eval when implemented
eval: false
---

The structure and content of the Data Package descriptor follow the
[Data Package standard](https://datapackage.org). This standard defines
the available properties at each level of the descriptor, specifies
which properties are required, and describes the allowed values for
each.

`check-datapackage` provides the `check()` function to verify that a
Data Package descriptor meets these requirements. `check()` takes a
Python `dict` as input, representing the descriptor, and by default
returns a list of `Issue` objects, each corresponding to a failed check
on a field in the descriptor. If you prefer the output as an error, you

> can also run `check()` in error mode. When checking a descriptor, you

Author: This looks a bit off to me because it creates the false impression that `check(error=True)` is correct code as is. But of course you need the descriptor as well. Maybe `check(..., error=True)`? Or they can scroll down to the section about this.

> can also run `check()` in error mode. When checking a descriptor, you

Author: ```suggestion
can also run `check()` with `error=True`. When checking a descriptor, you
```

can customise the function’s behaviour by specifying issues to ignore
and additional custom checks to perform.

## Default mode

> When you run `check()` in default mode, it returns a list of `Issue`

Reviewer: ```suggestion
When you run `check()`, by default it returns a list of `Issue`
```

If you want failed checks to result in errors and terminate program
execution, you can achieve this by setting the `error` argument of
`check()` to `True`.

```{python}
#| error: true
package_descriptor = {
    "name": 123,
    "title": "Hibernation Physiology of the Woolly Dormouse: A Scoping Review.",
    "id": "123-abc-123",
    "created": "2014-05-14T05:00:01+00:00",
    "version": "1.0.0",
    "licenses": [{"name": "odc-pddl"}],
}

cdp.check(descriptor=package_descriptor, error=True)
```

Now, if any checks fail, the function raises an error instead of
returning a list of issues.

## Configuration

You can pass a `Config` object to `check()` to customise the checks done
on the descriptor. The following configuration options are available:

-   `version`: The version of Data Package standard to check against.
    Defaults to `v2.`
-   `exclude`: The list of checks to exclude.
-   `rules`: The list of custom checks to run in addition to the checks
    defined in the standard.
-   `strict`: Whether to run recommended checks in addition to required
    ones. Defaults to `False`.

### Excluding checks

You can exclude checks based on their `type` and the fields they apply
to.

> The Data Package standard defines a range of check types (e.g.,

Author: Technically the types are defined by JSON schema and Data Package uses a subset of these

`path` field of **all** Data Resource descriptors:

```{python}
exclude_path = cdp.Exclude(target="$.resources[*].path")
```

The `type` and `target` arguments can also be combined:

```{python}
exclude_desc_required = cdp.Exclude(type="required", target="$.resources[*].description")
```

This would exclude required checks on the `description` field of Data
Resource descriptors.

To make the `check()` function aware of your exclusions, you add them to
the `Config` object passed to the function:

```{python}
package_descriptor = {
    "name": 123,
    "title": "Hibernation Physiology of the Woolly Dormouse: A Scoping Review.",
    "id": "123-abc-123",
    "created": "2014-05-14T05:00:01+00:00",
    "version": "1.0.0",
    "licenses": [{"name": "odc-pddl"}],
    "resources": [
        {
            "name": "woolly-dormice-2015",
            "title": "Body fat percentage in the hibernating woolly dormouse",
            "path": "https://en.wikipedia.org/wiki/Woolly_dormouse",
        }
    ],
}

config = cdp.Config(exclude=[exclude_required, exclude_name, exclude_path])
issues = cdp.check(descriptor=package_descriptor, config=config)
print(issues)
```

> In the example above, we would expect four checks to fail: the package

Reviewer: ```suggestion
In the example above, we would expect four `Issue` items: the package
```

To register your custom rules with the `check()` function, you add them
to the `Config` object passed to the function:

```{python}
package_descriptor = {
    "name": "woolly-dormice",
    "title": "Hibernation Physiology of the Woolly Dormouse: A Scoping Review.",
    "description": dedent("""
        This scoping review explores the hibernation physiology of the
        woolly dormouse, drawing on data collected over a 10-year period
        along the Taurus Mountain range in Turkey.
        """),
    "id": "123-abc-123",
    "created": "2014-05-14T05:00:01+00:00",
    "version": "1.0.0",
    "licenses": [{"name": "odc-pddl"}, {"name": "mit"}],
    "resources": [
        {
            "name": "woolly-dormice-2015",
            "title": "Body fat percentage in the hibernating woolly dormouse",
            "path": "resources/woolly-dormice-2015/data.parquet",
        }
    ],
}

config = cdp.Config(rules=[license_rule])
issues = cdp.check(descriptor=package_descriptor, config=config)
print(issues)
```

We can see that the custom rule was applied: `check()` returned one
issue flagging the first license attached to the Data Package.

> ### Strict mode

Reviewer: This section could stay here in the `check` guide

> The Data Package standard expresses both required and recommended rules.

Reviewer: ```suggestion
The Data Package standard has both required and recommended rules.
```

By default, `check()` checks only required rules. Recommended rules can
be turned on by setting the `strict` argument to `True`. The example
below violates the recommendation that the package `name` should contain
no special characters.

```{python}
package_descriptor = {
    "name": "Woolly Dormice (Toros Dağları)",
    "title": "Hibernation Physiology of the Woolly Dormouse: A Scoping Review.",
    "description": dedent("""
        This scoping review explores the hibernation physiology of the
        woolly dormouse, drawing on data collected over a 10-year period
        along the Taurus Mountain range in Turkey.
        """),
    "id": "123-abc-123",
    "created": "2014-05-14T05:00:01+00:00",
    "version": "1.0.0",
    "licenses": [{"name": "odc-pddl"}],
    "resources": [
        {
            "name": "woolly-dormice-2015",
            "title": "Body fat percentage in the hibernating woolly dormouse",
            "path": "resources/woolly-dormice-2015/data.parquet",
        }
    ],
}

issues = cdp.check(descriptor=package_descriptor, strict=True)
print(issues)
```

> ## Understanding `Issue`s

Reviewer: This could be it's own page

### The `Issue` class

One `Issue` corresponds to one failed check on one field in the
descriptor. An `Issue` has three attributes:

-   `type`: The type of the check that failed and caused the issue. This
    can be a type that exists in the Data Package standard (e.g.,
    `required` or `pattern`) or a custom type.
-   `location`: The location of the field that failed the check
    expressed in [JSON path](https://en.wikipedia.org/wiki/JSONPath)
    notation. For example, `$.resources[2].name`.
-   `message`: An explanation of what went wrong.

### The `explain()` function

The `explain()` function provides a more verbose and user-friendly
description of a list of `Issue`s. To use it, pass it the issues output
by `check()`:

```{python}
package_descriptor = {
    "name": 123,
    "title": "Hibernation Physiology of the Woolly Dormouse: A Scoping Review.",
    "id": "123-abc-123",
    "created": "2014-05-14T05:00:01+00:00",
    "version": "1.0.0",
    "licenses": [{"name": "odc-pddl"}],
}

issues = cdp.check(descriptor=package_descriptor)
print(cdp.explain(issues))

> ```

Reviewer: ````suggestion
```

When using `check(error=True)`, the `explain()` function will be used by default as
the error messages.
````

> ```

Author: ```suggestion
```

When setting `error=True` in `check()`, error messages will be generated
by the `explain()` function.
```
